### PR TITLE
Adding folder nodes mapping the Intercom hierarchy

### LINKED ISSUE #9158: [KW-Search] Update & backfill `folders` for intercom
See [#8527](https://github.com/dust-tt/dust/issues/8527)

### PR DESCRIPTION
**Pull request by overmode**
## Description
Add folder nodes for TeamId, Teams, CollectionId and HelpCenter

## Risk
Break Intercom connector

## Test
Connect locally, do a few operations, delete then restore connector.

## Deploy Plan
- Stop intercom connectors
- Deploy connectors
- Restart intercom connectors
- Run migration `20241219_backfill_intercom_data_source_folders.ts`

### CODE REVIEW

File: connectors/migrations/20241219_backfill_intercom_data_source_folders.ts [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+    const teams = await IntercomTeam.findAll({
+      where: {
+        connectorId: connector.id,
+      },
+    });
+    // Create a team folder for each team
+    await concurrentExecutor(
+      teams,
+      async (team) => {
+        const teamInternalId = getTeamsInternalId(connector.id);
```
Comment (aubin-tchoi) [inline connectors/migrations/20241219_backfill_intercom_data_source_folders.ts:44]: It's team without an S no?

File: connectors/migrations/20241219_backfill_intercom_data_source_folders.ts [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+        await upsertDataSourceFolder({
+          dataSourceConfig,
+          folderId: teamInternalId,
+          parents: [getTeamsInternalId(connector.id)],
```
Comment (aubin-tchoi) [inline connectors/migrations/20241219_backfill_intercom_data_source_folders.ts:48]: Should be the team followed by the teams folder (the one that contains all teams)

File: connectors/migrations/20241219_backfill_intercom_data_source_folders.ts [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+          collections,
+          async (collection) => {
+            const collectionInternalId = getHelpCenterCollectionInternalId(
+              connector.id,
+              collection.collectionId
+            );
+            await upsertDataSourceFolder({
+              dataSourceConfig,
+              folderId: collectionInternalId,
+              parents: [helpCenterInternalId],
```
Comment (aubin-tchoi) [inline connectors/migrations/20241219_backfill_intercom_data_source_folders.ts:104]: Don't you have to pass the folderId as well?

File: connectors/src/connectors/intercom/lib/conversation_permissions.ts [hunk 4]
```diff
@@ -51,6 +54,30 @@ export async function allowSyncTeam({
...
+      // Update datasource folder node
+      const connector = await ConnectorResource.fetchById(connectorId);
+      if (connector !== null) {
+        // Create folder node for the team
+        const teamInternalId = getTeamInternalId(
+          connectorId,
+          teamOnIntercom.id
+        );
+        await upsertDataSourceFolder({
+          dataSourceConfig: dataSourceConfigFromConnector(connector),
```
Comment (aubin-tchoi) [inline connectors/src/connectors/intercom/lib/conversation_permissions.ts:67]: I think we want to do these in the activity and not at a place where the user would wait for it

File: connectors/src/connectors/intercom/lib/conversation_permissions.ts [hunk 5]
```diff
@@ -51,6 +54,30 @@ export async function allowSyncTeam({
...
+          folderId: teamInternalId,
+          title: team.name,
+          parents: [teamInternalId, getTeamsInternalId(connectorId)],
+          mimeType: getDataSourceNodeMimeType("TEAM"),
+        });
+        await upsertDataSourceFolder({
+          dataSourceConfig: dataSourceConfigFromConnector(connector),
+          folderId: teamInternalId,
```
Comment (aubin-tchoi) [inline connectors/src/connectors/intercom/lib/conversation_permissions.ts:75]: I think there's a mixup with the above

File: connectors/src/connectors/intercom/lib/help_center_permissions.ts [hunk 6]
```diff
@@ -81,6 +85,23 @@ export async function allowSyncHelpCenter({
...
+
+      // Create datasource folder node
+      const connector = await ConnectorResource.fetchById(connectorId);
+      if (connector !== null) {
+        const dataSourceConfig = dataSourceConfigFromConnector(connector);
+        const helpCenterInternalId = getHelpCenterInternalId(
+          connectorId,
+          helpCenterOnIntercom.id
+        );
+        await upsertDataSourceFolder({
```
Comment (aubin-tchoi) [inline connectors/src/connectors/intercom/lib/help_center_permissions.ts:97]: Same as above, we probably want this in an activity

**Review by aubin-tchoi — COMMENTED** [review]
First batch of comments

File: connectors/src/connectors/intercom/lib/conversation_permissions.ts (hunk 4 above)
Comment (philipperolet) [reply connectors/src/connectors/intercom/lib/conversation_permissions.ts]: :+1:

File: connectors/src/connectors/intercom/lib/help_center_permissions.ts (hunk 6 above)
Comment (philipperolet) [reply connectors/src/connectors/intercom/lib/help_center_permissions.ts:97]: :+1:

**Review by philipperolet — COMMENTED** [review]
On the right path :+1:  Left some coms
Make sure to test 
Also checking: did you carefully track all model calls (to not miss any crud point)?

File: connectors/src/connectors/intercom/lib/utils.ts [hunk 7]
```diff
@@ -113,3 +132,78 @@ export function getConversationInAppUrl(
...
+}: {
+  documentId: string;
+  connectorId: number;
+  parentCollectionId: string;
+  helpCenterId: string;
+}) {
+  // Get collection parents
+  const collectionParents = await getParentIdsForCollection({
+    connectorId,
+    collectionId: parentCollectionId,
```
[the remark this answers is not in the document]
Comment (overmode) [reply connectors/src/connectors/intercom/lib/utils.ts:152]: Definitely agree, changing that

File: connectors/src/connectors/intercom/lib/help_center_permissions.ts [hunk 8]
```diff
@@ -220,6 +241,31 @@ export async function allowSyncCollection({
           getCollectionInAppUrl(intercomCollection, region),
         permission: "read",
       });
+
+      // Create datasource folder node
+      const connector = await ConnectorResource.fetchById(connectorId);
```
[the remark this answers is not in the document]
Comment (overmode) [reply connectors/src/connectors/intercom/lib/help_center_permissions.ts:246]: This one was actually redundant with the one within sync_help_center, deleting it

File: connectors/src/connectors/intercom/temporal/activities.ts [hunk 9]
```diff
@@ -481,6 +501,20 @@ export async function syncTeamOnlyActivity({
     name: teamOnIntercom.name,
     lastUpsertedTs: new Date(currentSyncMs),
   });
+
+  // Also make sure a datasource folder node is created for the team
+  if (connector !== null) {
```
Comment (aubin-tchoi) [inline connectors/src/connectors/intercom/temporal/activities.ts:506]: nit: you don't need this if

File: connectors/src/connectors/intercom/temporal/activities.ts [hunk 10]
```diff
@@ -698,3 +732,20 @@ export async function getSyncAllConversationsStatusActivity({
...
+}: {
+  connectorId: ModelId;
+}) {
+  const connector = await _getIntercomConnectorOrRaise(connectorId);
+  const dataSourceConfig = dataSourceConfigFromConnector(connector);
+
+  await upsertDataSourceFolder({
+    dataSourceConfig,
+    folderId: getTeamsInternalId(connectorId),
+    title: "Teams",
```
Comment (aubin-tchoi) [inline connectors/src/connectors/intercom/temporal/activities.ts:747]: This is called "Conversations" in the content node

File: connectors/src/connectors/intercom/temporal/activities.ts [hunk 11]
```diff
@@ -481,6 +501,20 @@ export async function syncTeamOnlyActivity({
     name: teamOnIntercom.name,
     lastUpsertedTs: new Date(currentSyncMs),
   });
+
+  // Also make sure a datasource folder node is created for the team
+  if (connector !== null) {
+    // Create folder node for the team
+    const teamInternalId = getTeamInternalId(connectorId, teamOnIntercom.id);
```
Comment (aubin-tchoi) [inline connectors/src/connectors/intercom/temporal/activities.ts:508]: TBH I'd rather have you use `teamOnDb` to have 1 way of doing this (same as when we get ContentNode)
if we think in Resources here, we would've had a way to get the parents, name and internalID from the Resource

File: connectors/src/connectors/intercom/temporal/activities.ts (hunk 9 above)
Comment (philipperolet) [inline connectors/src/connectors/intercom/temporal/activities.ts:506]: same here, throw if null :+1:

File: connectors/src/connectors/intercom/temporal/workflows.ts [hunk 12]
```diff
@@ -52,6 +55,11 @@ export async function intercomSyncWorkflow({
 }) {
   await saveIntercomConnectorStartSync({ connectorId });
 
+  // Add folder node for teams
+  await upsertIntercomTeamsFolderActivity({
+    connectorId,
+  });
+
```
Comment (aubin-tchoi) [inline connectors/src/connectors/intercom/temporal/workflows.ts:62]: we're in for a ride again @philipperolet :)

Comment (aubin-tchoi) [reply connectors/src/connectors/intercom/temporal/workflows.ts:62]: intercom is webhook-based for the conversations

Comment (aubin-tchoi) [reply connectors/src/connectors/intercom/temporal/workflows.ts:62]: oh wait actually we're good, I thiiink webhooks rely on non-temporal code

File: connectors/src/connectors/intercom/temporal/workflows.ts [hunk 13]
```diff
@@ -52,6 +55,11 @@ export async function intercomSyncWorkflow({
 }) {
   await saveIntercomConnectorStartSync({ connectorId });
 
+  // Add folder node for teams
+  await upsertIntercomTeamsFolderActivity({
```
Comment (philipperolet) [inline connectors/src/connectors/intercom/temporal/workflows.ts:59]: exceptionally, just this once, we could discuss moving that into the syncTeamOnly activity
To be discussed with @overmode @philipperolet @aubin-tchoi vs the difficulty of stopping / restarting all intercom workflows for this change

File: connectors/src/connectors/intercom/temporal/workflows.ts (hunk 12 above)
Comment (aubin-tchoi) [reply connectors/src/connectors/intercom/temporal/workflows.ts:62]: I would guess we can stop all the workflows, deploy and resume them, which will trigger full resyncs for all the Help Centers but should be ok

File: connectors/src/connectors/intercom/temporal/workflows.ts (hunk 13 above)
Comment (aubin-tchoi) [reply connectors/src/connectors/intercom/temporal/workflows.ts:59]: cc https://github.com/dust-tt/dust/pull/9469#discussion_r1893885581 I think we can get away with batch stop + batch resume

File: connectors/src/connectors/microsoft/lib/content_nodes.ts [hunk 14]
```diff
@@ -43,7 +43,7 @@ export function getTeamsRootAsContentNode(): ContentNode {
     }),
     parentInternalId: null,
     type: "folder",
-    title: "Teams",
+    title: "Conversations",
```
Comment (aubin-tchoi) [inline connectors/src/connectors/microsoft/lib/content_nodes.ts:46]: unintentional I would assume

File: connectors/src/connectors/intercom/temporal/workflows.ts (hunk 12 above)
Comment (overmode) [reply connectors/src/connectors/intercom/temporal/workflows.ts:62]: As far as I can tell the only worflow triggered by webhooks is stopIntercomSyncWorkflow -> Seems fine if we miss some

Comment (aubin-tchoi) [reply connectors/src/connectors/intercom/temporal/workflows.ts:62]: this workflow is triggered by a webhook on `/uninstall`, it's clearly not fine at all if we miss some but this one is more easily reconstructible (we need the connectorId in the webhook and we're good to go)

**Review by philipperolet — APPROVED** [review]

**Pull request merged**